src/network/networkd-varlink.c

```diff
@@ -1,6 +1,158 @@
+        _cleanup_free_ void *raw = NULL;
+        size_t l;
+        le64_t u;
+        int r;
+
+        assert(f);
+        assert(ret);
+
+        l = fread(&u, 1, sizeof(u), f);
+        if (l == 0 && feof(f))
+                return 0;
+        if (l != sizeof(u))
+                return -EBADMSG;
+
+        /* each LLDP packet is at most MTU size, but let's allow up to 4KiB just in case */
+        if (le64toh(u) >= 4096)
+                return -EBADMSG;
+
+        raw = new(uint8_t, le64toh(u));
+        if (!raw)
+                return -ENOMEM;
+
+        if (fread(raw, 1, le64toh(u), f) != le64toh(u))
+                return -EBADMSG;
+
+        r = sd_lldp_neighbor_from_raw(ret, raw, le64toh(u));
+        if (r < 0)
+                return r;
+
+        return 1;
+}
+
+static int dump_lldp_neighbors(int ifindex, Varlink *link) {
+        _cleanup_fclose_ FILE *f = NULL;
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+        int r;
+
+        assert(ifindex > 0);
+        assert(link);
+
+        r = open_lldp_neighbors(ifindex, &f);
+        if (r == -ENOENT) {
+                return varlink_replyb(link, JSON_BUILD_EMPTY_ARRAY);
+        }
+        if (r < 0)
+                return r;
+
+        for (;;) {
+                const char *chassis_id = NULL, *port_id = NULL, *system_name = NULL, *port_description = NULL;
+                _cleanup_(sd_lldp_neighbor_unrefp) sd_lldp_neighbor *n = NULL;
+                uint16_t cc;
+
+                r = next_lldp_neighbor(f, &n);
+                if (r < 0) {
+                        log_warning_errno(r, "Failed to read neighbor data: %m");
+                        break;
+                }
+                if (r == 0)
+                        break;
+
+                if (v) {
+                        r = varlink_notify(link, v);
+                        if (r < 0)
+                                return r;
+
+                        v = json_variant_unref(v);
+                }
+
+                (void) sd_lldp_neighbor_get_chassis_id_as_string(n, &chassis_id);
+                (void) sd_lldp_neighbor_get_port_id_as_string(n, &port_id);
+                (void) sd_lldp_neighbor_get_system_name(n, &system_name);
+                (void) sd_lldp_neighbor_get_port_description(n, &port_description);
+
+                r = sd_lldp_neighbor_get_enabled_capabilities(n, &cc);
+
+                r = json_build(&v, JSON_BUILD_OBJECT(
+                                JSON_BUILD_PAIR("ChassisId", JSON_BUILD_STRING(chassis_id)),
+                                JSON_BUILD_PAIR("SystemName", JSON_BUILD_STRING(system_name)),
+                                JSON_BUILD_PAIR_CONDITION(r >= 0, "EnabledCapabilities", JSON_BUILD_UNSIGNED(cc)),
+                                JSON_BUILD_PAIR("PortId", JSON_BUILD_STRING(port_id)),
+                                JSON_BUILD_PAIR_CONDITION(port_description, "PortDescription", JSON_BUILD_STRING(port_description))));
```

Comment:
I think it would be wise to mov this one level down, so that further parameters can be returned if needed. i.e. put this object in some field called "neighbor" or so inside some wrpping object that only has that field